internal/builders/docker/pkg/builder_test.go

```diff
@@ -157,13 +156,8 @@ func Test_Builder_SetUpBuildState(t *testing.T) {
 	}
 }
 
-// TODO(#1478): Use custom error types, and check errors based on their type.
-func checkError(t *testing.T, err error, want string) {
-	if err == nil {
-		t.Errorf("expected error, got nil")
-	}
-	got := fmt.Sprintf("%v", err)
-	if !strings.Contains(got, want) {
-		t.Errorf("unexpected error message: got (%q) does not contain want (%q)", got, want)
+func checkError(t *testing.T, got error, want any) {
+	if !errors.As(got, &want) {
```

Comment:
Unfortunately, `errors.As` in Go 1.19+ is a bit finicky. Right now this will pass for any input of `want`.
https://go.dev/play/p/hZcplmSC3lx

Unfortunately you need to implement a function per type you want to check as was done here
https://github.com/slsa-framework/slsa-github-generator/blob/4c2f963dd197da76843196f9830051d772d8efc2/internal/builders/generic/attest_test.go#L27-L32
See: https://go.dev/play/p/YuY_OIVNtcF

Reply:
Thanks for pointing this out. I have now made `checkError` generic. So now it fails if the wrong error type is used.

Reply:
That's a good idea. I should probably update attest_test.go to use this method.